language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -187,16 +204,21 @@ dummyType = TyVar () dummyTyName
 
 -- | Extract type of a term. The resulting type is normalized.
 typeOf :: Term TyNameWithKind NameWithType a -> TypeCheckM a (NormalizedType TyNameWithKind ())
-typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = pure (void $ NormalizedType ty) -- annotations on types must be normalized
-typeOf (LamAbs _ _ ty t)                         = NormalizedType <$> (TyFun () (void ty) <$> (getNormalizedType <$> typeOf t))
+typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = do
```

Comment:
Do we need annotations with type environments?

Reply:
How do you mean?

Reply:
I didn't spend much time studying about this algorithm, so I may ask silly questions. Here I attempted to ask why can't we track the types of term variables in type environments, but now I see that we only use type environments to avoid performing substitution. We could also use environments for term variables, but I guess there was a decision not to do so.



The question is "since we have type environments in which